contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,200 @@
+package tracing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"go.temporal.io/sdk/interceptor"
+	"go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/workflow"
+)
+
+var (
+	_ interceptor.Tracer     = tracerImpl{}
+	_ interceptor.TracerSpan = &tracerSpan{}
+)
+
+// NewInterceptor creates an interceptor for setting on client options
+// that implements Datadog tracing for workflows.
+func NewInterceptor() interceptor.Tracer {
+	return tracerImpl{}
+}
+
+type contextKey string
+
+const (
+	activeSpanContextKey contextKey = "dd_trace_span"
+	headerKey                       = string(activeSpanContextKey)
+)
+
+type tracerImpl struct {
+	interceptor.BaseTracer
+}
+
+func (tracerImpl) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: activeSpanContextKey,
+		HeaderKey:      headerKey,
+	}
+}
+
+func (tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func (tracerImpl) StartSpan(options *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	startOpts := []tracer.StartSpanOption{
+		tracer.ResourceName(options.Name),
+		tracer.StartTime(options.Time),
+	}
+
+	// Set a deterministic span ID for workflows which are long-running and cross process boundaries
+	if options.Operation == "RunWorkflow" {
+		rid, ok := options.Tags["temporalRunID"]
+		if !ok {
+			return nil, fmt.Errorf("missing required tag `temporalRunID` on `RunWorkflow` span")
+		}
+		h := fnv.New64()
+		// Write() always writes all bytes and never fails; the count and error result are for implementing io.Writer.
+		_, _ = h.Write([]byte(fmt.Sprintf("workflow:%s", rid)))
+		startOpts = append(startOpts, tracer.WithSpanID(h.Sum64()))
```

Comment:
We should add a test to ensure that the resulting span ID here is stable. It's a breaking change if the way we compute the span ID ever changes since child spans will become orphaned.

Reply:
Added a test, let me know what you think!